Drivers: Sensor: mc3419: Added LPF Configuration using DT properties

cx-anuj-pathak opened the pull request:
For applications that rely on low noise/variance in accelerometer sample reading. Application can take advantage of integrated LPF Thus this PR adds LPF configuration capability to the mc3419 driver

nordicjm reviewed (commented):
Why isn't this being set in device tree?

cx-anuj-pathak commented:
> Why isn't this being set in device tree?

Because of two reasons
- Since we already have `MC3419_DECIMATION_RATE` Kconfig variable that used to configure decimation rate. I don't want to confuse user with one configuration in Kconfig while other in dts. Thus `MC3419_LOW_PASS_FILTER_CUT*` was introduced
- not all values are valid for LPF settings, thus i find it is easy to restrict them while providing some info in help string, using Kconfig `choice`

gmarull reviewed (changes requested):
this belongs to DT

cx-anuj-pathak commented:
@gmarull @nordicjm 

1. Moving only LPF setting to DT, will have one configuration `LPF` in dts, while other `DECIMATION RATE` in Kconfig. 
2. Move Both LPF and DECIMATION RATE to DT, and update any existing sample.

Are you guys fine with 1st approach, or 2nd make more sense?

cx-anuj-pathak committed:
Drivers: Sensor: mc3419: Added LPF CutOff select using dt

For applications that rely on low noise/variance in accelerometer
sample reading. Application can take advantage of integrated LPF
Thus this PR adds LPF configuration capability to the mc3419 driver

Signed-off-by: Anuj Pathak <anuj@croxel.com>

cx-anuj-pathak committed:
Drivers: Sensor: mc3419: Moved Decimation Rate config to DT

- Similar to LPF settings decimation rate could be applied to
per instance basis, thus moved to DT prop

Signed-off-by: Anuj Pathak <anuj@croxel.com>

ubieda commented on dts/bindings/sensor/memsic,mc3419.yaml:
```diff
@@ -21,3 +21,30 @@ properties:
+      - 11
+      - 13
+    description: |
+      Select lpf cut-off frequency if lpf-en is true. otherwise this has no effect
+      Possible values and corresponding values are listed bellow
```
```suggestion
      Select lpf cut-off frequency if lpf-en is true. otherwise this has no effect
      Possible values and corresponding values are listed bellow.
      Default is disabling LPF.
```

ubieda commented on dts/bindings/sensor/memsic,mc3419.yaml:
```diff
@@ -21,3 +21,30 @@ properties:
+  decimation-rate:
+    type: int
+    default: 0
+    description: |
+      This helps in producing slower interrupt. Internal Data
+      Rate is divide by this decimation rate. If decimation rate
+      is 0 then internal data rate is equal to output data rate,
+      then it produce interrupt floods.
```
```suggestion
    description: |
      This helps in producing slower interrupt. Internal Data
      Rate is divide by this decimation rate. If decimation rate
      is 0 then internal data rate is equal to output data rate,
      then it produce interrupt floods.
      Default is disabling decimation.
```

jeppenodgaard commented on drivers/sensor/mc3419/mc3419.c:
```diff
@@ -298,9 +304,14 @@ static const struct sensor_driver_api mc3419_api = {
 #endif
 
 #define MC3419_DEFINE(idx)						\
+	BUILD_ASSERT(((DT_INST_PROP(idx, decimation_rate) >= 0) &&	\
```
I do not think it is possible to trigger this assert due to the enum values specified.

cx-anuj-pathak replied:
build assert is for `decimation-rate`, while `lpf-fc-sel` is enum

nordicjm commented:
> @gmarull @nordicjm
> 
>     1. Moving only LPF setting to DT, will have one configuration `LPF` in dts, while leave other `DECIMATION RATE` in Kconfig.
> 
>     2. Move Both LPF and DECIMATION RATE to DT, and update any existing sample.
> 
> 
> Are you guys fine with 1st approach, or 2nd make more sense?

Second makes sense, older drivers had options added to Kconfig but they never really should have been put there. Whilst an application with one of these devices is only ever likely to have one, there is no reason it can't have two or more, and if you have two or more and have the options set by Kconfig, that means every device has the exact same configuration and can't be changed, whereas with dts you can have as unique configuration as you want for each device

MaureenHelm commented on dts/bindings/sensor/memsic,mc3419.yaml:
```diff
@@ -21,3 +21,22 @@ properties:
+
+  lpf-fc-sel:
+    type: int
+    default: 0
```
Need to explain in the description why the default value was selected
https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#dt-bindings-default-rules

MaureenHelm commented on dts/bindings/sensor/memsic,mc3419.yaml:
```diff
@@ -40,3 +40,13 @@ properties:
+
+  decimation-rate:
+    type: int
+    default: 0
```
Need to explain in the description why the default value was selected
https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#dt-bindings-default-rules

MaureenHelm reviewed (commented):
Thanks for moving the decimation rate to devicetree

cx-anuj-pathak commented:
@gmarull friendly reminder

gmarull commented on dts/bindings/sensor/memsic,mc3419.yaml:
```diff
@@ -21,3 +21,47 @@ properties:
+      - 10
+      - 11
+      - 13
+    description: |
+      Select LPF cut-off frequency.
+      Possible values and corresponding values are listed below
+      0 => disable LPF
+      9 => LPF with fc set to IDR/4.255
+      10 => LPF with fc set to IDR/6
+      11 => LPF with fc set to IDR/12
+      13 => LPF with fc set to IDR/16
+      Default is LPF disabled (reset value of the register).
```
This is IMHO is a pretty bad DT usage. If I read in DT `lpf-fc-sel = <11>` I have no clue at all what's going on. Reading bindings isn't clear enough either, eg, what is IDR?

I'd expect something like:

- no property specified -> disable LPF
- `lpf-fc-scale` -> scale value

cx-anuj-pathak replied:
DT explanation is reworded, let me know if it more clear now. IDR was explained in decimation-rate property, i have included it in this too.
If no property is specified then i have to use DT_INST_PROP_OR with 0, isn't it more descriptive if we use 0 as default value in DT definition itself then silently using it?
And `lpf-fc-sel` are already explained as it is in the datasheet. Do you want me to add datasheet reference?

cx-anuj-pathak replied:
@gmarull let me know your thoughts on this too.

gmarull replied:
I guess then we need some sort of human readable defines on a header.

cx-anuj-pathak commented on drivers/sensor/mc3419/mc3419.c:
```diff
@@ -298,9 +304,14 @@ static const struct sensor_driver_api mc3419_api = {
 #define MC3419_DEFINE(idx)						\
+	BUILD_ASSERT(((DT_INST_PROP(idx, decimation_rate) >= 0) &&	\
+			(DT_INST_PROP(idx, decimation_rate) <= 15)),	\
+			"0 <= decimation rate <= 15. condition fails");	\
```
removed build assert and added used enum in DT definition.

cx-anuj-pathak commented on dts/bindings/sensor/memsic,mc3419.yaml:
```diff
@@ -21,3 +21,47 @@ properties:
+      Input Data Rate (IDR). Possible values and corresponding decimation
+      rate are listed below
+      0  => ODR = IDR
+      1  => ODR = IDR / 2
+      2  => ODR = IDR / 4
+      3  => ODR = IDR / 5
+      4  => ODR = IDR / 8
+      5  => ODR = IDR / 10
+      6  => ODR = IDR / 16
+      7  => ODR = IDR / 20
+      8  => ODR = IDR / 40
+      9  => ODR = IDR / 67
+      10 => ODR = IDR / 80
+      11 => ODR = IDR / 100
+      12 => ODR = IDR / 200
+      13 => ODR = IDR / 250
+      14 => ODR = IDR / 500
+      15 => ODR = IDR / 1000
```
I didn't get it. can you provide more details.
I'm new to open source contribution so forgive me if it is anything obvious.

gmarull replied:
you're forcing the users to introduce register values, not actual rate.

cx-anuj-pathak replied:
@gmarull that is because, i am unable to find pattern to easily convert real_decimation_rate to decimation_rate_cfg_value using a formula.
And if we still want to achieve that, we'll have to use a const array and trace it manually. that will have a unnecessary 16 length array to have ultimately use only single one, while rest 15 will be redundant. both code size and complexity will increase.

Let me know is there is a simple way, e.g. if it is possible to define enum variable, and use it's index in the code using some macro. Or something similar.
Otherwise i would argue a developer will relate it using dts documentation or datasheet.

cx-anuj-pathak replied:
@gmarull nevermind, i have found couple of references in dts/bindings/sensor with enum. and refactored it using a header file. please review

MaureenHelm commented on dts/bindings/sensor/memsic,mc3419.yaml:
```diff
@@ -21,3 +31,16 @@ properties:
+      Enable and select LPF cutoff frequency for a given IDR (Input Data Rate).
+      Default is LPF disabled (reset value of the register).
+    enum:
+      - 0  # MC3419_LPF_DISABLE
+      - 9  # MC3419_LPF_EN_WITH_IDR_BY_4p255_FC
+      - 10 # MC3419_LPF_EN_WITH_IDR_BY_6_FC
+      - 11 # MC3419_LPF_EN_WITH_IDR_BY_12_FC
+      - 13 # MC3419_LPF_EN_WITH_IDR_BY_16_FC
```
Move the comments describing the enumerations into the description. This will allow the content to be included in the generated documentation
